### PR TITLE
[AutoDiff] Emit a diagnostic for non-differentiable active values

### PR DESCRIPTION
**Pull request by asavonic**
For some values we cannot compute types for differentiation (for example, tangent vector type), so it is better to diagnose them earlier. Otherwise we hit assertions when generating code for such invalid values.

The LIT test is a reduced reproducer from the issue #66996. Before the patch the compiler crashed while trying to get a tangent vector type for the following value (partial_apply):

    %54 = function_ref @$s4null1o2ffSdAA1FV_tFSdyKXEfu0_ :
      $@convention(thin) @substituted <τ_0_0> (@inout_aliasable Double)
      -> (@out τ_0_0, @error any Error) for <Double>

    %55 = partial_apply [callee_guaranteed] %54(%2) :
      $@convention(thin) @substituted <τ_0_0> (@inout_aliasable Double)
      -> (@out τ_0_0, @error any Error) for <Double>

Now we emit a diagnostic instead.

The patch resolves issues #66996 and #63331.

### CODE REVIEW

**Comment by asl** [comment]
Tagging @BradLarson

**Comment by asl** [comment]
@swift-ci please test

File: lib/SILOptimizer/Differentiation/PullbackCloner.cpp [hunk 1]
```diff
@@ -1804,6 +1804,11 @@ bool PullbackCloner::Implementation::run() {
   DominanceOrder domOrder(original.getEntryBlock(), domInfo);
   // Keep track of visited values.
   SmallPtrSet<SILValue, 8> visited;
+  // Also check for differentiable for all active values
```
Comment (rxwei) [inline lib/SILOptimizer/Differentiation/PullbackCloner.cpp:1807]: Not sure I understand this comment. The two lines of code below are just obtaining the protocol and the module and don't seem related to the comment.

Comment (asavonic) [reply lib/SILOptimizer/Differentiation/PullbackCloner.cpp:1807]: Right, this comment is more suitable next to the actual check below.

**Comment by asl** [comment]
@swift-ci please test

**Comment by asavonic** [comment]
There is quite a lot of LIT test with tuples being incorrectly (?) diagnosed as non-differentiable. I will check them and update the patch.

**Comment by asl** [comment]
@asavonic Because some components are non-differentiable?

**Comment by asl** [comment]
@swift-ci please test

**Comment by asl** [comment]
@shahmishal any idea why CI is not triggered? (see https://github.com/apple/swift/pull/67697#issuecomment-1679437979)

**Review by rxwei — APPROVED** [review]
Sorry for the delayed review
